### PR TITLE
Make compiletest do exact matching on triples

### LINKED ISSUE #48893: Make compiletest look for exact matches in the arch table
#47737 reported a problem with architecture names that are substrings of each other, but the problem is more general than powerpc. #48732 removed the `powerpc` annotations, but [I argued afterwards](https://github.com/rust-lang/rust/issues/47737#issuecomment-370232447), it would be better to improve compiletest such that it doesn't consider e.g. `powerpc` to be a `powerpc64` target. In light of the arguments linked earlier and the additional fact that this *keeps happening* -- #48875 introduces the same problem for MIPS -- I really think we should revert #48732 and fix compiletest instead.

cc @nagisa @jcowgill @debris @alexcrichton

### PR DESCRIPTION
**Pull request by varkor**
This avoids the issues of the previous substring matching, ensuring `ARCH_TABLE` and `OS_TABLE` will no longer contain redundant entries. Fixes #48893.

r? @rkruppe

### CODE REVIEW

**Comment by sanxiyn** [comment]
I don't think this PR is correct, because we already rely on substring matching. For example, with this change, `arm-linux-androideabi` will not match `ignore-android`.

**Comment by varkor** [comment]
@sanxiyn: Ah, that's awkward — do you know how prevalent this is (that is: using substring matching)? Would it be plausible to special case strings like `android` instead?

**Comment by sanxiyn** [comment]
I think you should review the output of `rustc --print target-list` and add missing entries to `ARCH_TABLE` and `OS_TABLE`. It should be doable although tedious.

**Comment by rkruppe** [comment]
Putting all the minor variations of the triple into compiletest probably also makes adding new targets (incl. minor variations of existing targets) more annoying, but hopefully less so than learning months later your `ignore-*` lines don't do what they appear to be doing.

**Comment by varkor** [comment]
To clarify: you think the tests should be all updated too, so that:
`// ignore-android` becomes:
```
// ignore-android
// ignore-androideabi
```
I.e. no more substring matching whatsoever.

**Comment by rkruppe** [comment]
No, that's not what I'm saying. Triples contain more information (and in some cases, random noise) than we need for the purpose of "ignore this test on this architecture". For example, consider how `i386`, `i586` and `i686` triples are all considered for `ignore-x86` currently, or how `arm64` in triples is treated as an alias for `aarch64`. I'm in favor of extending that canonicalization to other variations in arch and OS components. This still requires someone to go through all the triples, but only in compiletest, not in every test ever (where it would add no useful information).

This is a judgement call to some degree, there may be distinctions in triples that we may want to preserve. But we'll learn about those when someone wants to make that distinction in a test.

**Comment by varkor** [comment]
Okay, I've added the OSes and architectures that weren't previously considered, and added what I think are reasonable canonicalisations for each of them — they should match the previous substring matching unless something very specific was being done before (which is probably most easily observed by letting bors test it).

File: src/tools/compiletest/src/util.rs [hunk 1]
```diff
@@ -14,43 +14,62 @@ use common::Config;
...
 const OS_TABLE: &'static [(&'static str, &'static str)] = &[
     ("android", "android"),
+    ("androideabi", "android"),
     ("bitrig", "bitrig"),
     ("cloudabi", "cloudabi"),
     ("darwin", "macos"),
     ("dragonfly", "dragonfly"),
+    ("eabi", "eabi"),
+    ("eabihf", "eabi"),
+    ("elf", "elf"),
```
Comment (rkruppe) [inline src/tools/compiletest/src/util.rs]: Uh these seem rather odd. I know triples can be weird sometimes, are these actual "operating systems" or do they just happen to be in that position in the triple?

Comment (rkruppe) [reply src/tools/compiletest/src/util.rs]: Um, I guess a better question is: which triples do there come from?

Comment (varkor) [reply src/tools/compiletest/src/util.rs]: They did seem odd, but I didn't feel certain enough to exclude them. They come from`thumbv6m-none-eabi` (and the other `thumb*` triples) and `msp430-none-elf`. Do you think I should just exclude these?

**Comment by varkor** [comment]
Not sure why bors thought I was trying anything funny... I know my place.

File: src/tools/compiletest/src/util.rs (hunk 1 above)
Comment (rkruppe) [reply src/tools/compiletest/src/util.rs]: Ah so these triples have no "vendor" component, so the *second* component is the "OS" and the last component is the *ABI*.

File: src/tools/compiletest/src/header.rs [hunk 2]
```diff
@@ -574,12 +574,14 @@ impl Config {
...
 
+            let triple: Vec<&str> = self.target.split('-').collect();
+
             name == "test" ||
-                util::matches_os(&self.target, name) ||             // target
-                name == util::get_arch(&self.target) ||             // architecture
-                name == util::get_pointer_width(&self.target) ||    // pointer width
-                name == self.stage_id.split('-').next().unwrap() || // stage
-                Some(name) == util::get_env(&self.target) ||        // env
+                util::matches_os(&self.target, triple.get(2), name) || // target
```
Comment (rkruppe) [inline src/tools/compiletest/src/header.rs]: I was always worried about "hard-coding" the meaning of triple components like this since they are somewhat flexible. Since it turns out we even have triples in-tree for which this breaks down, let's not do that...

File: src/tools/compiletest/src/util.rs (hunk 1 above)
Comment (varkor) [reply src/tools/compiletest/src/util.rs]: Ah — is it always safe to make this assumption? (I.e. elided component is vendor?)

File: src/tools/compiletest/src/util.rs [hunk 3]
```diff
@@ -14,70 +14,90 @@ use common::Config;
...
-            return os == name;
+pub fn matches_os(triple: &str, target_os: Option<&&str>, name: &str) -> bool {
+    if let Some(target_os) = target_os {
+        // For the wasm32 bare target we ignore anything also ignored on emscripten
+        // and then we also recognize `wasm32-bare` as the os for the target
+        if triple == "wasm32-unknown-unknown" {
+            return name == "emscripten" || name == "wasm32-bare"
+        }
+        for &(triple_os, os) in OS_TABLE {
+            if target_os == &triple_os {
```
Comment (rkruppe) [inline src/tools/compiletest/src/util.rs]: ... instead let's simply look at all triple components (like it previously did, just without the substring search now). Namespace collisions between the different triple components are practically impossible and would cause far worse problems in the ecosystem of "things that consume triples".

**Review by hanna-kruppe — CHANGES_REQUESTED** [review]

File: src/tools/compiletest/src/util.rs (hunk 3 above)
Comment (varkor) [reply src/tools/compiletest/src/util.rs]: Sure, I wasn't sure if the components could ever coïncide — but I shall do that instead.

**Comment by pietroalbini** [comment]
@varkor it was actually a bors bug, don't worry :)

File: src/tools/compiletest/src/header.rs [hunk 4]
```diff
@@ -574,12 +574,14 @@ impl Config {
...
 
+            let triple: Vec<&str> = self.target.split('-').collect();
+
             name == "test" ||
-                util::matches_os(&self.target, name) ||             // target
-                name == util::get_arch(&self.target) ||             // architecture
-                name == util::get_pointer_width(&self.target) ||    // pointer width
-                name == self.stage_id.split('-').next().unwrap() || // stage
-                Some(name) == util::get_env(&self.target) ||        // env
+                util::matches_os(&self.target, &triple, name) || // target
```
Comment (rkruppe) [inline src/tools/compiletest/src/header.rs]: Nit: I think now that these functions look at the whole triple again, there's not much to be gained from splitting here and passing the component explicitly (especially for `matches_os` which needs the whole triple anyway). Perhaps just revert these changes and split in the `util::*` functions to keep the interface simpler?

File: src/tools/compiletest/src/header.rs [hunk 5]
```diff
@@ -574,12 +574,14 @@ impl Config {
...
+                name == util::get_arch(&triple) ||               // architecture
+                name == util::get_pointer_width(&self.target) ||       // pointer width
+                name == self.stage_id.split('-').next().unwrap() ||    // stage
+                Some(&name) == triple.get(3) ||                        // env
```
Comment (rkruppe) [inline src/tools/compiletest/src/header.rs]: Nit: This is a pure refactoring, right? If so, I'd prefer to leave it out of this PR, for consistency. (Splitting multiple times is sure to have completely neglegible performance impact.)

**Review by hanna-kruppe — CHANGES_REQUESTED** [review]
I don't want to r- over this now, but I'd really appreciate if you squashed the commits after the revert because right now there's some back and forth (later commits undoing quite a bit of the earlier ones) which only makes sense in the context of this review, not when viewing history.

I also left two nits.

**Comment by varkor** [comment]
@rkruppe: bors can wait! I've squashed most of the commits down and simplified the changes.

**Comment by rkruppe** [comment]
Thanks! 

@bors r+
